File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/ui/collection/CollectionViewModel.kt
package com.jworks.kanjiquest.android.ui.collection

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jworks.kanjiquest.core.domain.model.CollectedItem
import com.jworks.kanjiquest.core.domain.model.CollectionItemType
import com.jworks.kanjiquest.core.domain.model.CollectionStats
import com.jworks.kanjiquest.core.domain.model.Rarity
import com.jworks.kanjiquest.core.domain.repository.CollectionRepository
import com.jworks.kanjiquest.core.domain.repository.KanjiRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class CollectionUiState(
    val selectedTab: CollectionItemType = CollectionItemType.KANJI,
    val items: List<CollectedItem> = emptyList(),
    val filteredItems: List<CollectedItem> = emptyList(),
    val stats: CollectionStats = CollectionStats(0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
    val selectedRarityFilter: Rarity? = null,
    val isLoading: Boolean = true,
    val kanjiLiterals: Map<Int, String> = emptyMap()
)

@HiltViewModel
class CollectionViewModel @Inject constructor(
    private val collectionRepository: CollectionRepository,
    private val kanjiRepository: KanjiRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(CollectionUiState())
    val uiState: StateFlow<CollectionUiState> = _uiState.asStateFlow()

    init {
        loadCollection()
    }

    private fun loadCollection() {
        viewModelScope.launch {
            val stats = collectionRepository.getCollectionStats()
            _uiState.value = _uiState.value.copy(stats = stats)
            selectTab(CollectionItemType.KANJI)
        }
    }

    fun selectTab(type: CollectionItemType) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, selectedTab = type)

            val items = collectionRepository.getCollectedByType(type)

            // Load kanji literals for display
            val literals = if (type == CollectionItemType.KANJI) {
                items.associate { item ->
                    val kanji = kanjiRepository.getKanjiById(item.itemId)
                    item.itemId to (kanji?.literal ?: "?")
                }
            } else _uiState.value.kanjiLiterals

            val filtered = applyFilter(items, _uiState.value.selectedRarityFilter)

            _uiState.value = _uiState.value.copy(
                items = items,
                filteredItems = filtered,
                kanjiLiterals = literals,
                isLoading = false
            )
        }
    }

    fun filterByRarity(rarity: Rarity?) {
        val filtered = applyFilter(_uiState.value.items, rarity)
        _uiState.value = _uiState.value.copy(
            selectedRarityFilter = rarity,
            filteredItems = filtered
        )
    }

    private fun applyFilter(items: List<CollectedItem>, rarity: Rarity?): List<CollectedItem> {
        return if (rarity != null) {
            items.filter { it.rarity == rarity }
        } else {
            items
        }
    }
}

File: ios-app/KanjiQuest/Screens/Collection/CollectionViewModel.swift
import Foundation
import SharedCore

enum CollectionTab: String, CaseIterable {
    case kanji = "Kanji"
    case hiragana = "Hiragana"
    case katakana = "Katakana"
    case radical = "Radicals"
}

@MainActor
final class CollectionViewModel: ObservableObject {
    @Published var selectedTab: CollectionTab = .kanji
    @Published var items: [CollectedItem] = []
    @Published var filteredItems: [CollectedItem] = []
    @Published var totalCollected = 0
    @Published var selectedRarityFilter: CollectedItemRarity? = nil
    @Published var kanjiLiterals: [String: String] = [:]
    @Published var isLoading = true

    private var collectionRepository: CollectionRepository?
    private var kanjiRepository: KanjiRepository?

    func load(container: AppContainer) {
        collectionRepository = container.collectionRepository
        kanjiRepository = container.kanjiRepository
        loadItems()
    }

    func selectTab(_ tab: CollectionTab) {
        selectedTab = tab
        loadItems()
    }

    func filterByRarity(_ rarity: CollectedItemRarity?) {
        selectedRarityFilter = rarity
        applyFilter()
    }

    private func loadItems() {
        Task {
            isLoading = true
            let type = selectedTab.rawValue.lowercased()
            items = (try? await collectionRepository?.getItemsByType(type: type)) ?? []
            totalCollected = items.count

            // Load kanji literals for kanji tab
            if selectedTab == .kanji {
                var literals: [String: String] = [:]
                for item in items {
                    if let kanji = try? await kanjiRepository?.getKanjiById(id: item.itemId) {
                        literals["\(item.itemId)"] = kanji.literal
                    }
                }
                kanjiLiterals = literals
            }

            applyFilter()
            isLoading = false
        }
    }

    private func applyFilter() {
        if let rarity = selectedRarityFilter {
            filteredItems = items.filter { $0.rarity == rarity }
        } else {
            filteredItems = items
        }
    }
}
